pkg/kubelet/mirror_manager_test.go

```diff
@@ -0,0 +1,214 @@
+	return &m
+}
+
+func (self *fakeMirrorManager) HasPod(podFullName string) bool {
+	self.mirrorPodLock.RLock()
+	defer self.mirrorPodLock.RUnlock()
+	return self.mirrorPods.Has(podFullName)
+}
+
+func (self *fakeMirrorManager) NumOfPods() int {
+	self.mirrorPodLock.RLock()
+	defer self.mirrorPodLock.RUnlock()
+	return self.mirrorPods.Len()
+}
+
+func (self *fakeMirrorManager) GetPods() []string {
+	self.mirrorPodLock.RLock()
+	defer self.mirrorPodLock.RUnlock()
+	return self.mirrorPods.List()
+}
+
+func (self *fakeMirrorManager) GetCounts(podFullName string) (int, int) {
+	self.mirrorPodLock.RLock()
+	defer self.mirrorPodLock.RUnlock()
+	return self.createCounts[podFullName], self.deleteCounts[podFullName]
+}
+
+// Tests that mirror pods are filtered out properly from the pod update.
+func TestFilterOutMirrorPods(t *testing.T) {
+	mirrorPod := api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			UID:       "987654321",
+			Name:      "bar",
+			Namespace: "default",
+			Annotations: map[string]string{
+				ConfigSourceAnnotationKey: "api",
+				ConfigMirrorAnnotationKey: "mirror",
+			},
+		},
+	}
+	staticPod := api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			UID:         "123456789",
+			Name:        "bar",
+			Namespace:   "default",
+			Annotations: map[string]string{ConfigSourceAnnotationKey: "file"},
+		},
+	}
+
+	expectedPods := []api.Pod{
+		{
+			ObjectMeta: api.ObjectMeta{
+				UID:         "999999999",
+				Name:        "taco",
+				Namespace:   "default",
+				Annotations: map[string]string{ConfigSourceAnnotationKey: "api"},
+			},
+		},
+		staticPod,
+	}
+	updates := append(expectedPods, mirrorPod)
+	actualPods, actualMirrorPods := filterAndCategorizePods(updates)
+	if !reflect.DeepEqual(expectedPods, actualPods) {
+		t.Errorf("expected %#v, got %#v", expectedPods, actualPods)
+	}
+	if !actualMirrorPods.Has(GetPodFullName(&mirrorPod)) {
+		t.Errorf("mirror pod is not recorded")
+	}
+}
+
+func TestParsePodFullName(t *testing.T) {
+	type nameTuple struct {
+		Name      string
+		Namespace string
+	}
+	successfulCases := map[string]nameTuple{
+		"bar_foo":         {Name: "bar", Namespace: "foo"},
+		"bar.org_foo.com": {Name: "bar.org", Namespace: "foo.com"},
+		"bar-bar_foo":     {Name: "bar-bar", Namespace: "foo"},
+	}
+	failedCases := []string{"barfoo", "bar_foo_foo", ""}
+
+	for podFullName, expected := range successfulCases {
+		name, namespace, err := ParsePodFullName(podFullName)
+		if err != nil {
+			t.Errorf("unexpected error when parsing the full name: %v", err)
+			continue
+		}
+		if name != expected.Name || namespace != expected.Namespace {
+			t.Errorf("expected name %q, namespace %q; got name %q, namespace %q",
+				expected.Name, expected.Namespace, name, namespace)
+		}
+	}
+	for _, podFullName := range failedCases {
+		_, _, err := ParsePodFullName(podFullName)
+		if err == nil {
+			t.Errorf("expected error when parsing the full name, got none")
+		}
+	}
+}
+
+func TestCreateMirrorPod(t *testing.T) {
```

Comment:
These last two tests I think should be in kubelet_test.go since they test Kubelet functionality. We can probably make the last one call garbageCollect directly if we'd like, but I do see the value of testing the Kubelet path which is why you wrote it this way.